### PR TITLE
Fgpu joint frontend

### PR DESCRIPTION
**Pull request by bmerry**
Checklist (if not applicable, edit to add `(N/A)` and mark as done):

- [x] If dependencies are added/removed: update `setup.cfg` and `.pre-commit-config.yaml`
- [x] (N/A) If modules are added/removed: use sphinx-apidoc to update files in `doc/`
- [x] Ensure copyright notices are present and up-to-date
- [x] (N/A) If qualification tests are changed: attach a sample qualification report
- [ ] If design has changed: ensure documentation is up to date
- [ ] If ICD-defined sensors have been added: update `fake_servers.py` in katsdpcontroller to match
- [ ] Update katsdpcontroller to pass only a single source
- [ ] Clarify how the digitiser ID polarisation bit works (NGC-1016)
- [ ] Update the ICD to indicate that the polarisations might be supplied in the right order.

As the checklist shows, there is still a bit of work to do, but this is already getting rather large.

Closes NGC-1006.

### CODE REVIEW

**Review by james-smith-za — COMMENTED** [review]
Overall it looks good to me. Just a few requests for clarification more than anything.

File: test/fgpu/test_postproc.py [hunk 1]
```diff
@@ -81,8 +81,7 @@ def postproc_host_pol(
 
 
 def postproc_host(
-    in0,
-    in1,
+    in_,
```
[the remark this answers is not in the document]
Comment (bmerry) [reply test/fgpu/test_postproc.py:84]: `in` is a keyword in Python. I could change it to something like `in_array` if you prefer.

File: src/katgpucbf/dsim/main.py [hunk 2]
```diff
@@ -161,7 +162,12 @@ async def async_main() -> None:
 
     timestamps = np.zeros(args.max_period // args.heap_samples, dtype=">u8")
     heap_sets = [
-        send.HeapSet.create(timestamps, [len(pol_dest) for pol_dest in args.dest], heap_size, range(len(args.dest)))
+        send.HeapSet.create(
+            timestamps,
+            [len(pol_dest) for pol_dest in args.dest],
+            heap_size,
+            range(args.first_id, args.first_id + len(args.dest)),
```
[the remark this answers is not in the document]
Comment (bmerry) [reply src/katgpucbf/dsim/main.py:169]: In run-dsim-hbw.sh, `len(args.dest)` is 1, so it will be either [0] or [1].

File: src/katgpucbf/fgpu/main.py [hunk 3]
```diff
@@ -425,14 +424,15 @@ def parse_args(arglist: Sequence[str] | None = None) -> argparse.Namespace:
...
+    parser.add_argument("src", type=parse_source, help="Source endpoints (or pcap file)")
     args = parser.parse_args(arglist)
 
     if args.use_peerdirect and not args.dst_ibv:
         parser.error("--use-peerdirect requires --dst-ibv")
-    for src in args.src:
-        if not isinstance(src, str) and args.src_interface is None:
-            parser.error("Live source requires --src-interface")
+    if not isinstance(args.src, str) and args.src_interface is None:
+        parser.error("Live source requires --src-interface")
```
[the remark this answers is not in the document]
Comment (bmerry) [reply src/katgpucbf/fgpu/main.py:433]: You still need to know which interface to send the IGMP join request to. Maybe unicast would work without `--src-interface` but I haven't made any attempt to support unicast.

File: src/katgpucbf/fgpu/engine.py [hunk 4]
```diff
@@ -1261,8 +1257,8 @@ def __init__(
 
     def _init_recv(self, src_affinity: list[int], monitor: Monitor) -> None:
         """Initialise the receive side of the engine."""
-        src_chunks_per_stream = 4
-        ringbuffer_capacity = src_chunks_per_stream * N_POLS
+        src_chunks = 4
```
[the remark this answers is not in the document]
Comment (bmerry) [reply src/katgpucbf/fgpu/engine.py:1260]: The chunks are twice as big now. So 4 chunks in one queue covers the same data that 4 chunks each in 2 queues used to.

File: src/katgpucbf/fgpu/kernels/pfb_fir.mako [hunk 5]
```diff
@@ -71,16 +72,34 @@ KERNEL REQD_WORK_GROUP_SIZE(WGS, 1, 1) void pfb_fir(
...
     int group_y = get_group_id(1);
+    int pol = get_group_id(2);
+    int in_offset;
+    switch (pol)
+    {
+% for pol in range(n_pols):
+    case ${pol}:
+        in_offset = in_offset${pol};
+        break;
+% endfor
```
[the remark this answers is not in the document]
Comment (bmerry) [reply src/katgpucbf/fgpu/kernels/pfb_fir.mako:99]: I'm not wild about it. I would have preferred to use an array of pointers, but that gets into a world of complication where you have to synchronise transferring new values into that array with running the kernel, and worrying that large HtoD transfers block that transfer.

File: test/fgpu/test_ddc.py [hunk 6]
```diff
@@ -31,34 +31,35 @@ def ddc_host(
...
-        (256, 16, 1234568, 13, None),
-        (256, 8, 1234568, 16, None),
-        (255, 4, 1234568, 32, None),
-        (257, 32, 1234568, 8, None),
-        (256, 64, 1234568, 5, None),
-        (32, 32, 1234568, 7, None),
-        (55, 5, 123464, 10, None),
-        (256, 16, 256 * 1024, 10, {"wgs": 96, "unroll": 4}),
+        (2, 256, 16, 256, 10, None),
+        (1, 256, 16, 1024 * 1024, 12, None),
```
[the remark this answers is not in the document]
Comment (bmerry) [reply test/fgpu/test_ddc.py:48]: Single-pol isn't used internally any more, but I was reluctant to remove support for it when it was easy enough to keep around. Maybe someone will adapt katgpucbf to build a single-pol F-engine one day (e.g. for very high bandwidths where one GPU can't handle both pols).

**Comment by bmerry** [comment]
I've pushed an update with the design doc changes.

File: test/fgpu/test_postproc.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (james-smith-za) [reply test/fgpu/test_postproc.py:84]: Ah yes of course.

No preference, I just didn't understand why you did it.

File: doc/engines.rst [hunk 7]
```diff
@@ -67,6 +67,13 @@ Stream
     This should not be confused with a CUDA stream, which corresponds to a
     Command Queue in OpenCL terminology.
 
+Stream Group
+    A group of incoming stress whose data are combined in chunks (see
```
Comment (james-smith-za) [inline doc/engines.rst:71]:
```suggestion
    A group of incoming streams whose data are combined in chunks (see
```

File: doc/engines.rst [hunk 8]
```diff
@@ -67,6 +67,13 @@ Stream
...
+    :class:`spead2.recv.ChunkStreamRingGroup`). Stream groups can be logically
+    treated like a single stream, and are used to scale receiving across
```
Comment (james-smith-za) [inline doc/engines.rst:73]: "are" used or "can be" used?
IIRC you're only giving the receive a single thread for the current case? Or did I misunderstand something?

File: doc/engines.rst (hunk 7 above)
Comment (bmerry) [reply doc/engines.rst:71]: I really need to stop using the wrong giraffes :-)

File: doc/engines.rst (hunk 8 above)
Comment (bmerry) [reply doc/engines.rst:73]: For high-bandwidth testing I'm using 4 streams. But let me try to reword it to better capture what I'm trying to say.

**Review by james-smith-za — APPROVED** [review]
